### PR TITLE
Deepseek v3

### LINKED ISSUE #11520: [Feature]: ValueError: Model architectures ['DeepseekV3ForCausalLM'] are not supported for now.
### 🚀 The feature, motivation and pitch

The DeepseekV3 is a very powerful model. Would you help support the model deployment?

https://huggingface.co/deepseek-ai/DeepSeek-V3

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by simon-mo**
Work in progress:
- [x] I still need to change quite a bit of code for FP8. Thanks @mgoin #11523
- [x] Need to run some eval to ensure the changes are correct. 
- [x] Turn off CUDA Graph for Deepseek V3 because the FP8 quant doesn't support it yet. 
- [ ] ~Kernel Tuning~ will do in a follow up

FIX #11520

```
vllm (pretrained=...,tensor_parallel_size=8,distributed_executor_backend=ray,trust_remote_code=true,max_model_len=4096,enforce_eager=true), gen_kwargs: (None), limit: 100.0, num_fewshot: 8, batch_size: 10
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  | 0.92|±  |0.0273|
|     |       |strict-match    |     8|exact_match|↑  | 0.92|±  |0.0273|
```

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -128,16 +136,17 @@ def forward_cpu(self, *args, **kwargs):
...
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        use_grouped_topk: bool,
+        top_k: int,
+        router_logits: torch.Tensor,
+        renormalize: bool,
+        topk_group: Optional[int] = None,
+        num_expert_group: Optional[int] = None,
+        custom_routing_function: Optional[Callable] = None,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/fused_moe/layer.py:148]: This should have the same signature as `forward_cuda`

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -73,16 +73,18 @@ def create_weights(self, layer: torch.nn.Module, num_experts: int,
...
-            custom_routing_function: Optional[Callable] = None
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        router_logits: torch.Tensor,
+        top_k: int,
+        renormalize: bool,
+        use_grouped_topk: bool,
+        topk_group: Optional[int] = None,
+        num_expert_group: Optional[int] = None,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/fused_moe/layer.py:84]: These need to be added to all methods that use MoE (gptq, compressed-tensors, etc)

File: vllm/model_executor/models/deepseek_v3.py [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+            final_hidden_states = final_hidden_states + shared_output
+        if self.tp_size > 1:
+            final_hidden_states = tensor_model_parallel_all_reduce(
+                final_hidden_states)
+
+        return final_hidden_states.view(num_tokens, hidden_dim)
+
+
+def yarn_get_mscale(scale: float = 1, mscale: float = 1) -> float:
+    import math
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/deepseek_v3.py:184]: move import to the top

File: vllm/model_executor/models/deepseek_v3.py [hunk 4]
```diff
@@ -0,0 +1,662 @@
...
+        if self.tp_size > config.n_routed_experts:
+            raise ValueError(
+                f"Tensor parallel size {self.tp_size} is greater than "
+                f"the number of experts {config.n_routed_experts}.")
+
+        if config.hidden_act != "silu":
+            raise ValueError(f"Unsupported activation: {config.hidden_act}. "
+                             "Only silu is supported for now.")
+
+        self.gate = MoEGate(config)
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/deepseek_v3.py:137]: use replicatedlinear

File: vllm/model_executor/models/deepseek_v3.py [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+            scaling_factor = rope_scaling["factor"]
+            mscale = yarn_get_mscale(scaling_factor, float(mscale_all_dim))
+            self.scaling = self.scaling * mscale * mscale
+
+        # self.attn = Attention(self.num_heads,
+        #                       self.qk_head_dim,
+        #                       self.scaling,
+        #                       num_kv_heads=self.num_heads)
+
+        # TODO, support head_size 192
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/deepseek_v3.py:286]: Can you add a better comment about what/why this is hardcoded for future self

File: vllm/model_executor/models/deepseek_v3.py [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+        )
+
+        # Fully Connected
+        hidden_states, residual = self.post_attention_layernorm(
+            hidden_states, residual)
+        hidden_states = self.mlp(hidden_states)
+        return hidden_states, residual
+
+
+@support_torch_compile
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/models/deepseek_v3.py:429]: ?

**Review by robertgshaw2-redhat — APPROVED** [review]
LGTM pending CUDAGraph enforce-eager

**Pull request merged**